File: Khushboo Singh/Array/Trapping Rain Water.java
class Solution {
    public int trap(int[] height) {
       int s=0,e=height.length-1,rm=0,lm=0,ans=0;
       while(s<e){
        if(height[s]<height[e]){
            if(height[s]>lm){
                lm=height[s++];

            }
            else{
                ans+=lm-height[s++];

            }
        }
        else{
            if(height[e]>rm){
                rm=height[e--];
            }
                
            else{
                ans+=rm-height[e--];
                }
            
            }
       }
       return ans;
    }
}

File: Abubakr Khanooni/Arrays/trappingRainWater.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n=height.size();
        vector<int>leftMax(height.size(),0);
        vector<int>rightMax(height.size(),0);

        leftMax[0]=height[0];
        rightMax[n-1]=height[n-1];
        for(int i=1;i<n;i++){
            leftMax[i]=max(height[i],leftMax[i-1]);
            rightMax[n-i-1]=max(height[n-i-1],rightMax[n-i]);
        }
        int count=0;
        for(int i=0;i<n;i++){
            count+=min(leftMax[i],rightMax[i])-height[i];
        }
        return count;
    }
};